accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+    a subtype of `Map<Object, Object>`.
+
+*   If a map spread element's static type implements `Map<K, V>` for some `K`
+    and `V` and `K` is not assignable to the key type of the map or `V` is not
+    assignable to the value type of the map.
+
+*   The type of the condition expression in an `if` element may not be assigned
+    to `bool`.
+
+    ```dart
+    [if ("not bool") 1] // Error.
+    ```
+
+*   The type of the iterator expression in a synchronous `for-in` element may
+    not be assigned to `Iterable<T>` for some type `T`. Otherwise, the *iterable
+    type* of the iterator is `T`.
+
+    ```dart
+    [for (var i in "not iterable") i] // Error.
+    ```
+
+*   The iterable type of the iterator in a synchronous `for-in` element may not
+    be assigned to the `for-in` variable's type.
+
+    ```dart
+    [for (int i in ["not", "int"]) i] // Error.
+    ```
+
+*   The type of the stream expression in an asynchronous `await for-in`
+    element may not be assigned to `Stream<T>` for some type `T`. Otherwise,
+    the *stream type* of the stream is `T`.
+
+    ```dart
+    [await for (var i in "not stream") i] // Error.
+    ```
+
+*   The stream type of the iterator in an asynchronous `await for-in` element
+    may not be assigned to the `for-in` variable's type.
+
+    ```dart
+    [await for (int i in Stream.fromIterable(["not", "int"])) i] // Error.
+    ```
+
+*   `await for` is used when the function immediately enclosing the collection
+    literal is not asynchronous.
+
+    ```dart
+    main() {
+      [await for (_ in stream)]; // Error.
+    }
+    ```
+
+*   `await` is used before a C-style `for` element. `await` can only be used
+    with `for-in` loops.
+
+    ```dart
+    main() {
+      [await for (;;)]; // Error.
+    }
+    ```
+
+*   The type of the condition expression (the second clause) in a C-style `for`
+    element may not be assigned to `bool`.
+
+    ```dart
+    [for (; "not bool";) 1] // Error.
+    ```
+
+## Constant Semantics
+
+The runtime semantics below can also be used to determine the compile-time value
+of a collection literal marked `const`, with a few modifications:
+
+*   A `listLiteral` is constant if it occurs in a constant context or if it is
+    directly prefixed by `const`. If so, then its `elements` must all be
+    constant elements, or at least potentially constant elements.
+
+*   A `setOrMapLiteral` is constant if it occurs in a constant context or if it
+    is directly prefixed by `const`. If so, then its `elements` must all be
+    constant elements, or at least potentially constant elements.
+
+*   An `expressionElement` is a constant element if its expression is a constant
+    expression, and a potentially constant element if it's expression is a
+    potentially constant expression.
+
+*   A `mapEntry` is a constant element if both its key and value expression are
+    constant expressions, and a potentially constant element if both are
+    potentially constant expressions.
```

Comment:
Same issue here, and in a few cases below: We may not want to have any potentially constant elements.

Reply:
We do want potentially const elements to support short-circuiting in `if`:

```dart
const dynamic d = null;
const list = [if (false) d + 3]; // OK.
```

The important part is that an if element can be constant (not potentially const) while still containing a potentially const element.